osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -107,9 +114,78 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
             // rate has to be applied each frame, because it's not guaranteed to be constant throughout playback
             // (see: ModTimeRamp)
-            drawableSpinner.Result.RateAdjustedRotation += (float)(Math.Abs(angle) * (gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate));
+            double time = gameplayClock?.CurrentTime ?? Clock.CurrentTime;
+            double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
+
+            currentDelta += rateAdjustedRotation;
+            drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();
+
+            Logger.Log(currentDelta.ToString(CultureInfo.InvariantCulture));
+        }
+
+        private float computeEffectiveRotation()
+        {
+            float curAngle = 0;
+            float maxAngle = 0;
+            float minAngle = 0;
+            float effectiveRotation = 0;
+
+            for (int i = 0; i < turningPoints.Count; i++)
+                applyDelta(turningPoints[i].delta);
+            applyDelta(currentDelta);
+
+            return effectiveRotation;
+
+            void applyDelta(float delta)
```

Comment:
I know you said this can be optimised, but I wasn't expecting this to run through every point every frame 😅

Reply:
Yeah I mean there is probably a simpler way to do this that I haven’t figured out yet.

Reply:
Well the simplest way would probably be extra memory, wouldn't it? Store (time, momentary effective rotation total) pairs in another list or something, and use the last momentary total to apply the delta onto. On rewind, pop the totals until stack only contains items in the past relative to the rewound position.

Don't have many more smarter ideas than that for now. If we only cared about time instants where spins are stored, we could only store that, but we need the instantaneous effective rotation values too for proper rewind playback.